### PR TITLE
Allow underscore in extended static files in theme

### LINKED ISSUE #32619: Underscores not allowed in module names in \Magento\Framework\App\Utility\Files::accumulateThemeStaticFiles
PSR-4 dictates that underscores have no special meaning and can be used in module names. There have been  [multiple](https://github.com/magento/magento2/pull/11765) [PRs](https://github.com/magento/magento2/pull/14397) accepted for this bug in other areas of Magento but this one has been missed.

This is in \Magento\Framework\App\Utility\Files::accumulateThemeStaticFiles. This method collates static files from your theme. The bug means that it's not possible to extend a static file from a module that has an underscore in the module name (eg. FishPig_WordPress_IntegratedSearch).

### Preconditions (*)
1. Magento 2.4.2 - but this is probably present in all versions from 2.0.0

### Steps to reproduce (*)
1. Install Magento 2.4.2
2. Extend a CSS file from a module with an underscore in the module name (app/design/frontend/YourVendor/YourTheme/ThirdParty_Module_Test/web/css/source/_module.less)
3. Compile the static assets

### Expected result (*)

The _module.less is processed correctly

### Actual result (*)
Error happened during deploy process: Could not parse theme static file '/var/www/test-siteapp/design/frontend/YourVendor/YourTheme/ThirdParty_Module_Test/web/css/source/_module.less'

---
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._

### PR DESCRIPTION
**Pull request by bentideswell**
This PR fixes [this issue](https://github.com/magento/magento2/issues/32619).

### Description
Adds the underscore character to the regex to allow for underscore in module name when extending static file in theme. This is inline with PSR-4 and also inline with other areas of Magento (see related pull requests).

### Manual Testing
1. Extend a file in your custom theme (original doesn't need to exist):
  app/design/frontend/YourVendor/YourTheme/Module_Third_Party/web/css/source/_module.less

2. Compile the static assets:

    bin/magento setup:static-content:deploy -f

You will now get the error:

Error happened during deploy process: Could not parse theme static file '/path/to/your/install/app/design/frontend/YourVendor/YourTheme/Module_Third_Party/web/css/source/_module.less'

If you then make the single character change in this PR and recompile, it will compile successfully.

### Related Pull Requests
[11765](https://github.com/magento/magento2/pull/11765)
[14397](https://github.com/magento/magento2/pull/14397)

### Fixed Issues
https://github.com/magento/magento2/issues/32619

### CODE REVIEW

**Comment by mrtuvn** [comment]
Should we allow such case ?
https://devdocs.magento.com/videos/fundamentals/create-a-new-module/
In my opinion named module with underscore may look confused. If we allow underscore i think we should update area related documents too

Can you describe how impact to user which this case ?

**Comment by bentideswell** [comment]
Magento already allows this. It is possible to create a module with an underscore in the name (I have several as a way to provide a basic sub-module system). The only issue is with static theme file collecting, the regex here forbids it. There have been other PRs accepted in the past for this issue in other areas ([11765](https://github.com/magento/magento2/pull/11765) & [14397](https://github.com/magento/magento2/pull/14397)) and both were quickly accepted.

Magento already allows this and PSR-4 allows this (which Magento claims to support).

By user, I meant more developer but the impact is that if a module is installed with an underscore in the module name that has a static file (I tested with view/frontend/web/css/source/_module.less) then this file cannot be extended in a custom theme due to the regex listed in the issue and PR I created for this issue.

This is a single character code change that brings the static theme file collecting inline with everywhere else in Magento.

**Comment by mrtuvn** [comment]
Thanks you for clarify! Can you cover this change with automatic tests ? (unit or integration would be great)

**Comment by bentideswell** [comment]
I have updated the original post with a better manual testing section.

I'm not sure how to cover this in unit testing, sorry!

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Static Tests, Integration Tests

**Comment by bentideswell** [comment]
@magento run Static Tests, Integration Tests

**Comment by bentideswell** [comment]
Integration test failed because of missing Magento copyright text but I’m not sure where to add it for that specific test. It seems unrelated to the issue though?

**Comment by mrtuvn** [comment]
That is false positive behavior from another file test and  may not related with your changes. Don't worry. Sometimes you need to re-run  some tests
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests CE, Functional Tests B2B, Integration Tests, WebAPI Tests

**Comment by bentideswell** [comment]
Is that another false positive? This test passed on the first run of tests yesterday. The Integration tests that failed last time has passed this time

**Comment by mrtuvn** [comment]
yes don't mind it! Core team will fix it for you

**Comment by bentideswell** [comment]
@magento run Functional tests

**Comment by bentideswell** [comment]
@magento run Functional Tests CE

File: dev/tests/integration/testsuite/Magento/Framework/App/Utility/FilesTest.php [hunk 1]
```diff
@@ -159,4 +159,16 @@ public function testReadLists()
...
+    /**
+     * Tests if underscores are allowed in extended static files in theme.
+     *
+     * @magentoComponentsDir Magento/Framework/App/Utility/_files/design
+     * @magentoAppIsolation enabled
+     * @magentoDbIsolation enabled
+     */
+    public function testAllowUnderscoreInExtendedFiles()
+    {
+        $this->model->getStaticPreProcessingFiles('*.less');
```
Comment (ihor-sviziev) [inline dev/tests/integration/testsuite/Magento/Framework/App/Utility/FilesTest.php]: @engcom-Foxtrot, I think we should make sure that we got the correct list of files and it contains file names with underscores. What do you think?

Comment (engcom-Foxtrot) [reply dev/tests/integration/testsuite/Magento/Framework/App/Utility/FilesTest.php]: @ihor-sviziev updated.

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

**Comment by ihor-sviziev** [comment]
@magento run WebAPI Tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-9071](https://jira.corp.magento.com/browse/ENGCOM-9071) has been created to process this Pull Request

**Comment by mrtuvn** [comment]
Hi just wonder is this issue can happen if theme name with 2 underscores ?
Example app/design/frontend/Magento/luma_custom_enhance/

**Comment by bentideswell** [comment]
Theme's aren't part of this issue, as far as I can tell, as the issue relates to PSR-4 autoloading of classes. With a theme name, there is no autoloading tied to the name of the theme as the theme doesn't create it's own PHP classes.

There may be some overlap though, especially in some of the regular expressions that are used to collect files that disallowed underscores. As this issue affected modules I created, I only debugged it from that perspective. If you have issues with a theme with underscores, you will need to:

1) find the places where it fails
2) test that it's actually failing because of the underscores
3) find the code causing the issue in the core
4) provide a fix or ask on here for help (I'm happy to take a look if you find the problem code)

All of this being said, I had no choice but to push to have the issue resolved as it affected modules my business had already sold and people were using. If it's a custom theme you're building, it might be easier to just change the name to something that works (if the underscore truly is your issue) and move on. I don't know enough about your situation but when I went through this, if it was possible for me to just remove the underscores without affecting many live sites, I would have definitely considered it!

**Pull request merged**